File: game/snake.rb
require_relative 'game'
require_relative 'board'

class Snake
  attr_reader :board
  attr_reader :data

  def initialize(board, data)
    @board = board
    @data = data
  end

  %w[id name health body].each do |key|
    define_method("#{key}_data") { data[key] }
  end

  alias id id_data
  alias name name_data
  alias health health_data

  def you?
    @you ||= self == board.game.you
  end

  def ==(other)
    id == other.id
  end

  def body
    @body ||= body_data.map { |body| Segment.new(self, body['x'], body['y']) }
  end
  alias segments body

  def head
    body.first
  end

  def enemy?
    !you?
  end

  def killable?
    enemy? && board.game.you.size > segments.size
  end

  def index
    @index ||= board.snakes.index { |snake| snake.id == id }
  end

  def inspect
    "<#{you? ? 'You' : 'Snake'} id=#{id} name=#{name} health=#{health} x=#{head.position.x} y=#{head.position.y}>"
  end
end

class Segment
  attr_reader :snake
  attr_reader :position

  def initialize(snake, x, y)
    @snake = snake
    @position = snake.board[x, y]

    position.segments << self
  end

  def head?
    self == snake.head
  end

  def inspect
    "<Segment snake_id=#{snake.id} x=#{position.x} y=#{position.y}>"
  end
end

File: game/snake.go
package game

type Snake struct {
	ID     string    `json:"id"`
	Name   string    `json:"name"`
	Health int       `json:"health"`
	Body   []Segment `json:"body"`

	board *Board
}

func (s *Snake) init(board *Board) {
	s.board = board

	for i := range s.Body {
		s.Body[i].init(s)
	}
}

func (s *Snake) Equals(other *Snake) bool {
	return s.ID == other.ID
}

func (s *Snake) Length() int {
	return len(s.Body)
}

func (s *Snake) Head() Segment {
	return s.Body[0]
}

func (s *Snake) OnBody(p Position) bool {
	for _, segment := range s.Body {
		if segment.Position.Equals(p) {
			return true
		}
	}

	return false
}

func (s *Snake) IsHeadAt(p Position) bool {
	return s.Head().Position.Equals(p)
}

func (s *Snake) IsYou() bool {
	return s.Equals(s.board.state.You)
}

type Segment struct {
	Position
	Snake *Snake
}

func (s *Segment) init(snake *Snake) {
	s.Snake = snake
}

func (s *Segment) Equals(other Segment) bool {
	return s.Snake.Equals(other.Snake) && s.Position.Equals(other.Position)
}

func (s *Segment) IsHead() bool {
	return s.Equals(s.Snake.Head())
}
